backend/src/v2/compiler/argocompiler/argo.go

```diff
@@ -180,69 +184,96 @@ func (c *workflowCompiler) templateName(componentName string) string {
+// referenced by the workflow templates using AWF templating syntax. The reason
+// for storing it in the workflow arguments is because there is a 1-many
+// relationship between components and tasks that reference them. The workflow
+// arguments allow us to deduplicate the component logic (implementation & spec
+// in IR), significantly reducing the size of the argo workflow manifest.
+func (c *workflowCompiler) saveProtoToArguments(componentName string, msg proto.Message) error {
 	if c == nil {
 		return fmt.Errorf("compiler is nil")
 	}
-	if c.wf.Annotations == nil {
-		c.wf.Annotations = make(map[string]string)
+	if c.wf.Spec.Arguments.Parameters == nil {
+		c.wf.Spec.Arguments = wfapi.Arguments{Parameters: []wfapi.Parameter{}}
 	}
-	if _, alreadyExists := c.wf.Annotations[name]; alreadyExists {
-		return fmt.Errorf("annotation %q already exists", name)
+	if c.wf.Spec.Arguments.GetParameterByName(componentName) != nil {
+		return nil
 	}
 	json, err := stablyMarshalJSON(msg)
 	if err != nil {
-		return fmt.Errorf("saving component spec of %q to annotations: %w", name, err)
+		return fmt.Errorf("saving component spec of %q to arguments: %w", componentName, err)
 	}
-	// TODO(Bobgy): verify name adheres to Kubernetes annotation restrictions: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set
-	c.wf.Annotations[name] = json
+	c.wf.Spec.Arguments.Parameters = append(c.wf.Spec.Arguments.Parameters, wfapi.Parameter{
+		Name:  componentName,
+		Value: wfapi.AnyStringPtr(json),
+	})
 	return nil
 }
 
-func (c *workflowCompiler) annotationPlaceholder(name string) (string, error) {
+// argumentsPlaceholder checks for the unique component name within the workflow
+// arguments and returns a template tag that references the component in the
+// workflow arguments.
+func (c *workflowCompiler) argumentsPlaceholder(componentName string) (string, error) {
 	if c == nil {
 		return "", fmt.Errorf("compiler is nil")
 	}
-	if _, exists := c.wf.Annotations[name]; !exists {
-		return "", fmt.Errorf("using component spec: failed to find annotation %q", name)
+	if c.wf.Spec.Arguments.GetParameterByName(componentName) == nil {
+		return "", fmt.Errorf("using component spec: failed to find workflow parameter %q", componentName)
 	}
-	// Reference: https://argoproj.github.io/argo-workflows/variables/
-	return fmt.Sprintf("{{workflow.annotations.%s}}", name), nil
+
+	return workflowParameter(componentName), nil
+}
+
+// extractBaseComponentName removes the iteration suffix that the IR compiler
+// adds to the component name.
+func ExtractBaseComponentName(componentName string) string {
+	baseComponentName := componentName
+	componentNameArray := strings.Split(componentName, "-")
+
+	if _, err := strconv.Atoi(componentNameArray[len(componentNameArray)-1]); err == nil {
+		baseComponentName = strings.Join(componentNameArray[:len(componentNameArray)-1], "-")
+
+	}
+
+	return baseComponentName
 }
```

Comment:
If I recall correctly, component function names that have underscores also get converted to dashes `-` in the IR `.components` section, example: 

```python
@dsl.component()
def foo_1(name: str) -> str:
    hello_text = f"Hello 2, {name}!"
    print(hello_text)
    return hello_text
```

would result in: 

```yaml
components:
  comp-foo-1:
```

In this case it seems like this function would use the baseComponentName `comp-foo`, which seems incorrect no?

Another alternative could be to introduce a separate sdk field that keeps the component name in check, maybe [here](https://github.com/kubeflow/pipelines/blob/master/api/v2alpha1/pipeline_spec.proto#L88), but then we have to consider backwards incompatibility.

Reply:
Yup! Nice catch! This is an edge case that we were thinking of tackling in a discrete PR, since the component naming logic already gets confused -- if you have a component named `foo` and a component named `foo_2` and you invoke `foo` twice, you wind up with components named `comp-foo`, `comp-foo-2`, and `comp-foo-2-2` in the IR. We can include that edge case handling in this PR though. The easiest way to do that would be to raise an exception in the SDK when a component function name culminates in `_<int(s)>`. Any backward compatibility concerns would only take place at the IR compilation stage, which (a) happens client-side and is easily resolved, and (b) would be an exceedingly rare edge case. Open to alternative approaches ofc!

Reply:
>  if you have a component named foo and a component named foo_2 and you invoke foo twice, you wind up with components named comp-foo, comp-foo-2, and comp-foo-2-2 

Yeah in this case it looks like it would extract the right base names for each. The edge case I mentioned will occur every time you have components with unique base names with a `_<int>` prefix.

The current approach feels a little hacky to me, because we are trying to rebuild info that should have been provided by the sdk initially. For example, in the alternative I proposed we could have sdk provide something like: 

```yaml
components:
  comp-foo-1:
    componentName: foo
    executorLabel: exec-foo-1
  comp-foo-1-1:
    componentName: foo-1
    executorLabel: exec-foo-1-2
```
And just use the component name `componentName`, instead of inferring it. Due to backwards incompatibility this would require we support this only future KFP versions though.

Anyways, food for thought. It is an edge case as you mentioned that is likely very rare. 

> We can include that edge case handling in this PR though. The easiest way to do that would be to raise an exception in the SDK when a component function name culminates in _<int(s)>.  

I'm personally fine with this, we should provide documentations on component naming maybe [here](https://www.kubeflow.org/docs/components/pipelines/concepts/component/).

I would also suggest adding some simple unit tests to illustrate the handling of the different component name formats.

Reply:
I would like to avoid introducing naming constraints if possible. Introducing a naming constraints is a breaking changes. 
The current state of appending component name with index suffix isn't ideal, but that's more of an implementation detail which can be changed without user impact, and we can dedupe them: https://github.com/kubeflow/pipelines/blob/eea7d483793c2b4f8c3aa20b1500535f7fbf50d4/sdk/python/kfp/deprecated/dsl/_component_bridge.py#L675

Reply:
Copy that. Thanks, @chensun!

One solution could be to update [ExtractBaseComponentName](https://github.com/kubeflow/pipelines/pull/10979/files#diff-481aa8f29d58257fef3e48fb60fd320f73d50b858018fe6adf7ac583125fc2c6R267) to return the full component name if the IR includes a task with a name in the format of `.*-\d+-\d+`. If you think this makes sense, maybe @zazulam and I can crank it out tomorrow morning before the community call. Just have to confirm with @zazulam about bandwidth. 

Another approach could  preserving dash `-` characters in the IR instead of converting them into underscores `_`. That way we could use `-<index>` in `make_name_unique_by_adding_index` and not have to worry about collisions since python function names cannot include dashes. Sounds easy in theory but maybe including dashes violates expectations in the protobuf / backend.

Reply:
> The current state of appending component name with index suffix isn't ideal, but that's more of an implementation detail which can be changed without user impact, and we can dedupe them: 

if we don't address it here, and we fear this breaking change can impact users, then we still need to resolve this as it would introduce a regression for this niche case. @droctothorpe correct me if I'm wrong here, it's been a couple of weeks since I looked at this, but in this case the component can assume the wrong parameters if paired with another component incorrectly.

> One solution could be to update [ExtractBaseComponentName](https://github.com/kubeflow/pipelines/pull/10979/files#diff-481aa8f29d58257fef3e48fb60fd320f73d50b858018fe6adf7ac583125fc2c6R267) to return the full component name if the IR includes a task with a name in the format of .*-\d+-\d+.

so I understand, this sounds like the alternative I proposed [here yes](https://github.com/kubeflow/pipelines/pull/10979#discussion_r1672251048)? If so I like this approach more. We can add it so as not to introduce any regression, and only support this for future sdk/kfp releases.

Reply:
You mean [this](https://github.com/kubeflow/pipelines/pull/10979#discussion_r1672595012) specific suggestion, right, @HumairAK? i.e. add component name to the IR spec to eliminate any guess work. I think that might be the ideal solution. Having an explicit map from the task name to the component name is an important piece of information to store in the IR. The only drawback that I can think of is that it theoretically increases the complexity of this PR, but if it's the desired target state, then the complexity is justified. @zazulam can you please mark this conversation as unresolved? Thanks!

Reply:
>The only drawback that I can think of is that it theoretically increases the complexity of this PR, but if it's the desired target state, then the complexity is justified.

Yeah, this is why I proposed sticking with what we had, given that it does introduces a backwards incompatible change, but for a very niche case, and the fix is relatively straight forward. 

But, if there is concerns around the impact, going IR spec route is the only way I can think of that is least destructive. I suppose we could also preserve the dashes, I'm not super familiar with the reasoning behind why we didn't initially, so my concern there would be possible around unforeseen side effects.

Reply:
Thanks for confirming, @HumairAK.

To summarize the three proposed solutions:

1. Update [ExtractBaseComponentName](https://github.com/kubeflow/pipelines/pull/10979/files#diff-481aa8f29d58257fef3e48fb60fd320f73d50b858018fe6adf7ac583125fc2c6R267) to return the full component name if the IR includes a task with a name in the format of .*-\d+-\d+.
2. Preserve dash `-` characters in the IR instead of converting them into underscores `_`. That way we could use `-<index>` in `make_name_unique_by_adding_index` and not have to worry about collisions since python function names cannot include dashes. Sounds easy in theory but including dashes may violate expectations in the protobuf / backend.
3. Add an explicit `componentName` field to each component in the IR so that the backend doesn't have to deduce the "base name."

Do you have a preference between these three approaches, @chensun? Are there other options that we haven't taken into consideration?

Reply:
Implemented and validated a solution. Continuing the thread [here](https://github.com/kubeflow/pipelines/pull/10979#issuecomment-2263795625) because this is too dang nested.